cEP-0020.md

```diff
@@ -0,0 +1,167 @@
+
+![](https://image.ibb.co/gJrdB7/datamodel.png)
+
+### Tracking Activities
+
+After we have the data, we need to track some important activities of newcomers like:
+
+- Created Issues
+- Created Pull Requests
+- Number of Gitmate errors on the PR
+- CI status
+- Review Comments
+- Issue Comments
+- Number of PRs Reviewed
+- Activities on Gitter channles
+- Introduce `.coafile` in other projects
+
+Which will help us in showing useful metrics for newcomers on a dashboard.
+
+#### Code Implementation
+
+Assuming that we are using [django-trackstats](https://github.com/pennersr/django-trackstats) for tracking the statics of newcomers, then we will make different domains of similar set of metrics like for activities related to pull requests, we will have a domain named PullRequest. i.e:
+
+```
+from trackstats.models import Domain
+Domain.objects.PullRequest = Domain.objects.register(
+    ref='pullrequest',
+    name='pullrequest')
+```
+
+And then we will define defferent metics to track in each of the domain:
+
+```
+Metric.objects.PULL_REQUEST_COUNT = Metric.objects.register(
+    domain=Domain.objects.PullRequest,
+    ref='pull_request_count',
+    name='Number of pull requests opened by the user')
+```
+
+Now we can store these metrics for the time period we want:
+
+```
+from trackstats.models import StatisticByDate, Domain, Metric, Period
+n = Order.objects.all().count()
+StatisticByDate.objects.record(
+    metric=Metric.objects.PULL_REQUEST_COUNT,
+    value=n,
+    Period=Period.LIFETIME)
+```
+
+### Gamification
+
+As we have already discuss the game elements involed in the gamification system, we will assign some points to each of the activities we will be tracking for a newcomer and then accordingly we will unlock the levels and provide them badges. now assuming the implementation of this process with [django-gamification](https://github.com/mattjegan/django-gamification) package, we need to link our `user` model to the gamification interface:
+
+```
+from .models import User
+from django.db import models
+from django_gamification.models import GamificationInterface
+
+class UserModel(models.User):
+    interface = models.ForeignKey(GamificationInterface)
+```
+
+Then we can create our first level:
+
+```
+from django_gamification.models import UnlockableDefinition
+
+UnlockableDefinition.objects.create(
+    name='Beginner',
+    description='You’re a beginner at coala',
+    points_required=50
+)
+```
+
+Similarly, we can create badges and award the user accordingly:
+
+```
+from django_gamification.models import BadgeDefinition, Category
+
+BadgeDefinition.objects.create(
+    name='Developer',
+    description='You’re now developer at coala',
+    points=300,
+    progression_target=100,
+    category=Category.objects.create(name='Developer Badges', description='This is the best bagde'),
+)
+```
+
+```
+from django_gamification.models import Badge
+
+badge = Badge.objects.first()
+badge.award()
+```
```

Comment:
not clear, where will the newcomers be able to see these points and badges. thought of being set them beside their name in gitter like some actual game XP sounds cool.

Reply:
We will make a dashboard in the community repo.

Reply:
tweak ``corobo`` to fetch those details, maybe in the form of a leaderboard :laughing:

Reply:
Will consider it in future to add this functionality in corobo, but for now, I think its good to have the dashboard on a page.

Reply:
Would it be a good idea to add stuff like top commiters , top reviewers of the month/week in the community repo?

Reply:
Sounds cool. :)

Reply:
Note that for this project, those stats pages would need to be limited to newcomers only.

reviewers data will come from the meta-review project -- reviews needs to be weighted (meta-review) to be useful, otherwise it encourages stupid reviews.

I think top commits can be obtained via the OpenHub data, however that can be a bit old, making it out of sync with reality, and thus not very good for weekly or even monthly stats.  Still useful for leaderboard for newcomers, who generally progress quite slowly anyway.

Needs to include the affiliated and unaffiliated committers:

https://www.openhub.net/orgs/coala/?view=affiliated_committers
https://www.openhub.net/orgs/coala/outside_committers

For weekly & monthly, better to use 'merged PRs' as a rough substitute for commits.  Possibly include lines added/removed/modified to weight PRs.